nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedRouteRecord.java

```diff
@@ -0,0 +1,177 @@
+import java.util.stream.Collectors;
+
+@Tags({"record", "routing", "script", "groovy", "jython", "python", "segment", "split", "group", "organize"})
+@CapabilityDescription(
+        "This processor provides the ability to route the records of the incoming FlowFile using an user-provided script. " +
+        "The script is expected to handle a record as argument and return with a string value. " +
+        "The returned value defines a route. All routes are bounded to an outgoing relationship where the record will be transferred to. " +
+        "Relationships are defined as dynamic properties: dynamic property names are serving as the name of the route. " +
+        "The value of a dynamic property defines the relationship the given record will be routed into. Multiple routes might point to the same relationship. " +
+        "Creation of these dynamic relationship is managed by the processor. " +
+        "The records, which for the script returned with an unknown relationship name are routed to the \"unmatched\" relationship. " +
+        "The records are batched: for an incoming FlowFile, all the records routed towards a given relationship are batched into one single FlowFile."
+)
+@SeeAlso(classNames = {
+    "org.apache.nifi.processors.script.ScriptedTransformRecord",
+    "org.apache.nifi.processors.script.ScriptedPartitionRecord",
+    "org.apache.nifi.processors.script.ScriptedValidateRecord",
+    "org.apache.nifi.processors.script.ScriptedFilterRecord"
+})
+@DynamicRelationship(name = "Name from Dynamic Property", description = "FlowFiles that match the Dynamic Property's Attribute Expression Language")
+public class ScriptedRouteRecord extends ScriptedRouterProcessor<String> {
+
+    static final Relationship RELATIONSHIP_ORIGINAL = new Relationship.Builder()
+            .name("original")
+            .description(
+                "After successful procession, the incoming FlowFile will be transferred to this relationship. " +
+                "This happens regardless the records are matching to a relationship or not.")
+            .build();
+
+    static final Relationship RELATIONSHIP_FAILURE = new Relationship.Builder()
+            .name("failed")
+            .description("In case of any issue during processing the incoming FlowFile, the incoming FlowFile will be routed to this relationship.")
+            .build();
+
+    static final Relationship RELATIONSHIP_UNMATCHED = new Relationship.Builder()
+            .name("unmatched")
+            .description("Records where the script evaluation returns with an unknown partition are routed to this relationship.")
+            .build();
+
+    private static Set<Relationship> RELATIONSHIPS = new HashSet<>();
+
+    static {
+        RELATIONSHIPS.add(RELATIONSHIP_ORIGINAL);
+        RELATIONSHIPS.add(RELATIONSHIP_FAILURE);
+        RELATIONSHIPS.add(RELATIONSHIP_UNMATCHED);
+    }
+
+    private final AtomicReference<Set<Relationship>> relationships = new AtomicReference<>();
```

Comment:
Why do we exactly need this? Do you think it'd worth renaming it to something like _atomicRelationships_?

Reply:
This processor provides the possibility to dynamically add routes, contrary to most of the processors. As processor instances might be called from multiple threads, we need to ensure thread safety. For your second question: I do not see the value of the renaming.

Reply:
The value is to differentiate _relationships_ from _RELATIONSHIPS_. From their name you cannot see the difference.